### ISSUE TITLE
Many threads causes reading of target index to be slow?

### ISSUE THREAD

**Issue by dkoslicki** [mahmudhera/smash_alternate_implementations#45]
Curious observation: using 25 threads as in the following:
```
dmk333@e5-cse-cbdmk02:/scratch/dmk333_new/temp/Salmonella/sourmash_alternate_implementations/bin$ ./compare /scratch/dmk333_new/temp/Salmonella/data/target_genomes/ABLENI010000000.fasta.with_name.sig.filelist /scratch/dmk333_new/temp/Salmonella/data/indexed /scratch/dmk333_new/temp/Salmonella/data/temp_working_dir /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes3.csv -c 0.001 -k 31 -t 25
**************************************
*
*   Query filelist: /scratch/dmk333_new/temp/Salmonella/data/target_genomes/ABLENI010000000.fasta.with_name.sig.filelist
*   Targets index directory: /scratch/dmk333_new/temp/Salmonella/data/indexed
*   Working directory: /scratch/dmk333_new/temp/Salmonella/data/temp_working_dir
*   Output filename: /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes3.csv
*   Containment threshold: 0.001
*   Number of threads: 25
*   Number of hash tables: 4096
*   Number of passes: 1
*   K-mer size: 31
*
**************************************
Reading all sketches using 25 threads
Number of query sketches to read: 1
Query sketches reading completed in 0 seconds.
Reading the target index...
Target index loaded in 29 seconds.
Computing all v all containment values...
Using 1 threads for pass 1/1
Computation progress: 100.00%
Writing progress: 100.00%
Pass 1/1 done.
Containment values computed in 27 seconds.
Writing the results to /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes3.csv
Results written to /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes3.csv
Removing the individual files in the working directory...
Individual files removed.
Cleaning up and exiting... (may take some time)
```
With an elapsed wall clock time of 1min 12 seconds.

Whereas the same command with 128 threads takes ~20x longer to read the index:
```
dmk333@e5-cse-cbdmk02:/scratch/dmk333_new/temp/Salmonella/sourmash_alternate_implementations/bin$ ./compare /scratch/dmk333_new/temp/Salmonella/data/target_genomes/ABLENI010000000.fasta.with_name.sig.filelist /scratch/dmk333_new/temp/Salmonella/data/indexed /scratch/dmk333_new/temp/Salmonella/data/temp_working_dir /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes2.csv -c 0.001 -k 31 -t 128
**************************************
*
*   Query filelist: /scratch/dmk333_new/temp/Salmonella/data/target_genomes/ABLENI010000000.fasta.with_name.sig.filelist
*   Targets index directory: /scratch/dmk333_new/temp/Salmonella/data/indexed
*   Working directory: /scratch/dmk333_new/temp/Salmonella/data/temp_working_dir
*   Output filename: /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes2.csv
*   Containment threshold: 0.001
*   Number of threads: 128
*   Number of hash tables: 4096
*   Number of passes: 1
*   K-mer size: 31
*
**************************************
Reading all sketches using 128 threads
Number of query sketches to read: 1
Query sketches reading completed in 0 seconds.
Reading the target index...
Target index loaded in 498 seconds.
Computing all v all containment values...
Using 1 threads for pass 1/1
Computation progress: 100.00%
Writing progress: 100.00%
Pass 1/1 done.
Containment values computed in 33 seconds.
Writing the results to /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes2.csv
Results written to /scratch/dmk333_new/temp/Salmonella/data/ani_results/threshold_0.001_with_sketch_sizes/threshold_0.001_with_sketch_sizes2.csv
Removing the individual files in the working directory...
Individual files removed.
Cleaning up and exiting... (may take some time)
```
with an elapsed wall clock time of 15min 27.40 seconds.

No immediate action requested, but something to keep in mind (perhaps a test) for when serialized formats like capnproto are considered.

**Comment by mahmudhera** [comment]
I think I understand the reason behind the behavior, I can make a fix now so the second code snippet that you gave does not happen (hopefully). But definitely will keep it in mind.

**Comment by mahmudhera** [comment]
Resolved by #46
